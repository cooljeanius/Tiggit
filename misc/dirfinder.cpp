#include "dirfinder.hpp"

#include <boost/filesystem.hpp>
#include <stdlib.h>
#include <fstream>

using namespace Misc;
using namespace boost::filesystem;

#ifdef _WIN32

#include <windows.h>
#include <shlobj.h>
#include <tchar.h>

TCHAR pathbuf[MAX_PATH];

static std::string getPathCSIDL(int csidl)
{
  SHGetFolderPath(NULL, csidl, NULL, 0, pathbuf);
  return std::string(pathbuf);
}

std::string DirFinder::getAppData()
{
  return getPathCSIDL(CSIDL_LOCAL_APPDATA);
}

std::string DirFinder::getExePath()
{
  char buf[2000];
  GetModuleFileName(NULL, buf, 2000);
  return std::string(buf);
}

static std::string getDataDir(int csidl, const std::string &vname, const std::string &aname)
{
  path p = getPathCSIDL(csidl);
  if(vname != "") p /= vname;
  return (p / aname).string();
}

bool DirFinder::getStandardPath(std::string &dir)
{
  dir = getDataDir(CSIDL_LOCAL_APPDATA, vname, aname);
  return isWritable(dir);
}

static bool openRegKey(const std::string &vname, const std::string &aname, HKEY &hkey)
{
  std::string where = "Software\\" + vname + "\\" + aname;
  LPCTSTR sk = where.c_str();

  LONG openRes = RegCreateKeyEx(HKEY_CURRENT_USER,
                                sk,
                                0,NULL,
				REG_OPTION_NON_VOLATILE,
				KEY_ALL_ACCESS,NULL,
                                &hkey,
                                NULL);

  return openRes == ERROR_SUCCESS;
}


static bool os_setStoredPath(const std::string &dir,
                             const std::string &vname,
                             const std::string &aname,
                             const std::string &dname)
{
  HKEY hkey;
  if(!openRegKey(vname,aname,hkey))
    return false;

  DWORD dwSize = dir.size();
  if(RegSetValueEx(hkey, dname.c_str(), 0, REG_SZ,
                   (const BYTE*)dir.c_str(), dwSize) != ERROR_SUCCESS)
    return false;

  RegCloseKey(hkey);
  return true;
}

static std::string os_getStoredPath(const std::string &vname,
                                    const std::string &aname,
                                    const std::string &dname)
{
  HKEY hkey;
  if(!openRegKey(vname,aname,hkey))
    return "";

  DWORD dwType = REG_SZ;
  DWORD dwSize = MAX_PATH;
  if(RegQueryValueEx(hkey, dname.c_str(), 0, &dwType, 
                     (BYTE*)pathbuf, &dwSize) != ERROR_SUCCESS)
    return "";

  RegCloseKey(hkey);

  return std::string(pathbuf);
}

#else
#ifdef __unix

static std::string getHome(const std::string &aname)
{
  /* More and more systems and apps are using ~/.local/share/appname
     instead of ~/.appname now. And I think that's a much less messy
     solution.
   */
  return std::string(getenv("HOME")) + "/.local/share/" + aname + "/";
}

static std::string getPathFile(const std::string &aname,
                               const std::string &dname)
{
  return getHome(aname) + dname + ".conf";
}

std::string DirFinder::getAppData() { return ""; }

bool DirFinder::getStandardPath(std::string &dir)
{
  dir = getHome(aname);
  return isWritable(dir);
}

std::string DirFinder::getExePath() { return ""; }

static std::string os_getStoredPath(const std::string &vname,
                                    const std::string &aname,
                                    const std::string &dname)
{
  std::string file = getPathFile(aname,dname);
  if(exists(file))
    {
      std::ifstream inp(file.c_str());
      if(inp)
        {
          std::string res;
          inp >> res;
          if(inp)
            return res;
        }
    }
  return "";
}

static bool os_setStoredPath(const std::string &dir,
                             const std::string &vname,
                             const std::string &aname,
                             const std::string &dname)
{
  std::string file = getPathFile(aname,dname);
  create_directories(path(file).parent_path());
  std::ofstream out(file.c_str());
  if(out)
    {
      out << dir;
      out.flush();
      if(out)
        return true;
    }
  return false;
}

#else
#error "This platform is not supported yet."
#endif
#endif

bool DirFinder::getStoredPath(std::string &dir)
{
  dir = os_getStoredPath(vname,aname,dname);
  return isWritable(dir);
}

bool DirFinder::setStoredPath(const std::string &dir)
{
  return isWritable(dir) && os_setStoredPath(dir,vname,aname,dname);
}

bool DirFinder::isWritable(const std::string &dir)
{
  if(dir == "")
    return false;

  try
    {
      // Create directories
      create_directories(dir);

      // Do write check
      path p = dir;
      p /= "xyz.tmp";
      std::string file = p.string();
      {
        std::ofstream out(file.c_str());
        if(!out) return false;
        out << "abc";
        if(!out) return false;
      }
      {
        std::ifstream inp(file.c_str());
        if(!inp) return false;
        std::string tmp;
        inp >> tmp;
        if(!inp || tmp != "abc")
          return false;
      }

      remove(p);
    }
  // All errors means the dir is not writable
  catch(...) { return false; }
  return true;
}
