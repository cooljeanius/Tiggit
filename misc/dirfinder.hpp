#ifndef __DIRFINDER_HPP_
#define __DIRFINDER_HPP_

#include <string>

/*
  This class finds an appropriate writable data path for the
  application.

  It can propose platform-dependent standard paths, or accept a
  user-provided value. It will remember settings between sessions.

  We only accept directories that are writable, and we test dirs for
  writability.

  You can have multiple data directories and store independent
  settings for each, if you provide different names to the
  constructor. The default name is 'default'.

  NOTE: On Windows Vista and newer, UAC virtualization will render
  write tests useless for determining access rights. It's highly
  recommended to disable virtualization by including a manifest file
  with your application, and specifing your privileges as "asInvoker"
  or similar.
  More info:
      http://msdn.microsoft.com/en-us/library/bb756960.aspx
      http://msdn.microsoft.com/en-us/library/bb756929.aspx

  Usage:

  A good practice is to call getStoredPath() first, and ALWAYS prompt
  the user if it returns false. The path returned by getStandardPath()
  may then be presented to the user as a default value.

  Whatever the user supplies, return it to DirFinder through
  setStoredPath().
 */

namespace Misc
{
  struct DirFinder
  {
    /* The vendor_name, app_name and dirname should be filesystem and
       windows registry compatible names. Preferably short, lower-case
       and without spaces.
     */
    DirFinder(const std::string &vendor_name,
              const std::string &app_name,
              const std::string &dirname = "default")
      : vname(vendor_name), aname(app_name), dname("path-" + dirname) {}

    /* Get stored path. Returns true if found and writable, false
       otherwise.

       If a path is set but the return value is false, it means the
       stored path was invalid or not writable.
     */
    bool getStoredPath(std::string &path);

    /* Get OS-dependent standard path for applications. Will only
       return true if an appropriate writable path was found.

       This does NOT change the stored path, but it may create the
       given path if it does not exist. Use setStoredPath() to use
       this path.
     */
    bool getStandardPath(std::string &path);

    /* Windows only. Returns CSIDL_LOCAL_APPDATA folder. Returns an
       empty string on other platforms.
     */
    static std::string getAppData();

    /* Get current executable path. Currently windows only.
     */
    static std::string getExePath();

    /* Set stored path. Will return true if the path was accepted,
       ie. it was writable, and we were able to store the information
       in a retrievable way.

       If the path was accepted, future requests to getStoredPath()
       will return this path.
     */
    bool setStoredPath(const std::string &path);

    /* Tests if a given directory is valid and writable as a data
       path. Will create directories if they don't exist.
     */
    static bool isWritable(const std::string &path);

  private:
    std::string vname,aname,dname;
  };
}

#endif
