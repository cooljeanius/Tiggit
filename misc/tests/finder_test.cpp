#include <iostream>
#include <fstream>
using namespace std;

#include "dirfinder.hpp"

Misc::DirFinder fnd("tiggit.net", "tiggit", "finder-test");

void setTest(const string &dir)
{
  cout << "Attempting to set stored path to '" << dir << "': ";
  if(fnd.setStoredPath(dir))
    cout << "SUCCESS\n";
  else
    cout << "FAILURE\n";
}

int main()
{
  string out;

  if(fnd.getStoredPath(out))
    cout << "Found stored path: ";
  else
    cout << "No stored path found, out = '";
  cout << out << (fnd.isWritable(out)?"' (ok)":"' (not ok)") << endl;

  if(fnd.getStandardPath(out))
    cout << "Standard path acceptible: ";
  else
    cout << "Standard path NOT acceptible: ";
  cout << out << (fnd.isWritable(out)?" (ok)":" (not ok)") << endl;

  setTest("/blah");
  setTest(out);
#ifdef _WIN32
  setTest("c:\\windows\\system\\");
#else
  setTest("/");
#endif

  if(fnd.getStoredPath(out))  
    cout << "Current stored path: " << out << endl;
  else
    cout << "FAILED getting stored path (out=" << out << ")\n";
  return 0;
}
