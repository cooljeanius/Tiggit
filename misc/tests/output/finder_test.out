Found stored path: /home/mortennk/.local/share/tiggit/' (ok)
Standard path acceptible: /home/mortennk/.local/share/tiggit/ (ok)
Attempting to set stored path to '/blah': FAILURE
Attempting to set stored path to '/home/mortennk/.local/share/tiggit/': SUCCESS
Attempting to set stored path to '/': FAILURE
Current stored path: /home/mortennk/.local/share/tiggit/
